Refactor to be fully LSP server-side

JustinGrote opened the pull request:
- [x] Replaced Custom LSP calls with OmniSharp `IPrepareRenameHandler` and `IRenameHandler`
- [x] Use custom `ScriptExtentAdapter` and `ScriptPositionAdapter` types to translate between script and lsp positions to avoid "off-by-one" errors since lsp is zero-based and script is one-based
- [x]  Tests for PrepareRenameProvider and RenameProvider
- [x]  Deduplicate PrepareRenameProvider and RenameProvider logic
- [x] Add Configuration Support via Omnisharp ILanguageServerConfiguration
- [x] Create Settings for Function and Variable Aliases and[ pass thru the info to LSP](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#workspace_configuration)
- [ ] Wire up Settings to actually work 
- [x] Add Message Support via ILanguageServerFacade
- [x] Present Risk Warning and acceptance via LSP `ShowMessageRequest` (cannot set the setting directly ATM as it's not an LSP standard scenario)
- [x] Rewrite IterativeFunctionVisitor to an `AstVisitor`
- [x] Rewrite IterativeVariableVisitor to an `AstVisitor`
- [x] Finish fixing tests
- [x] Add more "sad path" tests for things that should not get triggered for rename
- [x] Add E2E tests for testing the actual rename process
- [x] Validate FunctionName and VariableName to be valid before passing through
- [ ] Support scope modifiers (e.g. $GLOBAL, $SCRIPT, etc.)

Potential Additional "Aliasing Feature"
- [ ] Renaming $_ or $PSItem can place a $_ = or $PSItem = at the top of the scope. (I would use this a lot and seems pretty reasonable to do, but we would need an implicit definition of $PSItem in the current structure since we require a definition)

Potential Additional Features
Simple conditional splat parameter rename.
```
function Do-Thing($test) {} #test should rename
$x = @{test = 2} #test here should rename
$x.test = 5 #Test here should rename
Do-Thing @x
```

More Tests
- [x] Space in variable or function rename request should fail "$va riable" along with any other invalid characters for a rename
- [ ] Parameter, Function, and Variable renames where they are not defined in scope should fail
- [x] Renaming a parameter and removing the '-' should still work

JustinGrote committed:
Move all Handling to OmniSharp LSP types

JustinGrote committed:
Move HandlerError

JustinGrote committed:
Rework initial AST filter

JustinGrote committed:
Reorganize Tests under Handler folder

JustinGrote committed:
Lots of removing of custom types. Currently broken until I create a ScriptExtent/Position Adapter

JustinGrote committed:
Rework RenameHandler to use Adapters

JustinGrote committed:
Fix namespacing

JustinGrote committed:
Remove Alias from tests for now, will have separate Alias test in future

JustinGrote committed:
Default CreateAlias to false per feedback

JustinGrote committed:
Reworked Visitor to use ScriptPositionAdapter

JustinGrote committed:
Fixup tests with default PowerShell Formatting

JustinGrote committed:
Refine VariableVisitor to use ScriptExtentAdapter

JustinGrote committed:
Extract some duplicate functions and enable NRT checking for RenameHandler

JustinGrote committed:
Add initial EUA prompt scaffolding

JustinGrote committed:
Move RenameHandler to TextDocument (more appropriate context)

JustinGrote committed:
Remove Unnecessary code

JustinGrote committed:
Split out RenameService. **TESTS NEED FIXING**

JustinGrote committed:
Introduce service to the Extension

JustinGrote committed:
Redo tests to be data-driven

JustinGrote committed:
Fixup Tests, still a bug in rename logic with functions

JustinGrote committed:
Fixed all function Prepare tests

JustinGrote committed:
Fixed all variable PrepareRenameHandler Tests

JustinGrote committed:
First Stage work to move to a more stateless AstVisitor for renames

JustinGrote committed:
Separate out AstExtensions and continue Functions Reimplement. FunctionsSingle test works at least

JustinGrote committed:
Move renameservice

JustinGrote committed:
Breakout and simplify matching logic, fixes more test cases

JustinGrote committed:
SkipChildren doesn't work for nested function situations. More tests fixed

JustinGrote committed:
All rename function tests fixed

JustinGrote committed:
Add disclaimer scaffolding (needs config)

JustinGrote committed:
Add more config setup and lock down some classes

JustinGrote committed:
Add configuration and adjust opt-in message due to server-side LSP config limitation

JustinGrote committed:
Add mocks for configuration for testing, still need to fix some parameter tests

JustinGrote committed:
Bonus Assertions

JustinGrote commented:
@Razmo99 I see you haven't been active for a few months so if I don't hear from you in the next week what I'll do is force commit over the top of your PR so that you get primary author credit. Thanks so much for your hard work, your variable visitor itself is pretty impressive! I mostly just had to rewrite it to be 100% LSP server side so that other platforms such as NeoVim can take advantage too.

Razmo99 commented:
@JustinGrote  apologies been busy with some life things, but I've seen the changes you've been making it looks like it's coming along great, thank you for your work on this aswell. Happy for you to go ahead with what u are planning, I'm also interested to see what you changed to make it LSP I've learned alotnworking with you this will have been my first collaboration on a public project

JustinGrote commented:
@Razmo99 You can take a look at this branch, I've made a *lot* of changes but primarily:
- Make it 100% server side, I mostly just followed the LSP spec and use the types in omnisharp, I'm no means an expert.
- Similarly, replace all the custom stuff you made with omnisharp/LSP/PowerShell types where possible to cut down on the amount of "original" code here. There may come a day we move away from omnisharp as an LSP but we can innersource the data object types at that point and probably be OK.
- I rewrote your function renamer to an AstVisitor and took out a lot of the state and made it more flow-functional, because we can make a fair assumption that it will use the default Visitor flow which is depth-first so we don't need to track what we have alreay processed.
- I rewrote the test scaffolding to take out a lot of boilerplate duplication, and moved most of the tests out to the `Handler` level, since everything else is technically an implementation detail and the `Handler` is our public API, having the tests here mean we can change how the internals work and still have a good suite.
- The variable visitor is awesome, great work here, I'll probably mostly preserve it but may rewrite it an AstVisitor if possible.

I'll update the notes up top. Again, thanks so much for your hard work on this and solving a lot of problems, all I'm doing is making it fit some of the style of the project.

Razmo99 commented:
@JustinGrote sounds awsome! I'll checkout that branch sometime today to see the changes and how it works now. It sounds like overall it'll be alot cleaner and easier to develop further, can't wait to get this officially released

JustinGrote committed:
Fix all Tests

JustinGrote committed:
Actually fix tests (missing some pattern matching on AST types)

JustinGrote committed:
Small format adjust

JustinGrote committed:
Format Update

JustinGrote committed:
Move IterativeVariableVisitor into RenameService class

JustinGrote committed:
Remove unnecessary intermediate tests

JustinGrote committed:
Internalize rename visitor into renameservice

JustinGrote committed:
Perform initial abstraction of visitors to a base class

JustinGrote committed:
Add disclaimer link

JustinGrote committed:
Apply formatting to test fixtures

JustinGrote committed:
Fix issue with dependency injection weirdly setting Disclaimer to true and fix when dialog is closed rather than just button

JustinGrote committed:
Change name of Alias Setting

JustinGrote committed:
Make the PrepareRenameSymbol return more legible

JustinGrote committed:
Add support for negative test cases

JustinGrote committed:
I thought I fixed all these...

JustinGrote committed:
Small test fixes and naming updates

JustinGrote committed:
Move AstExtensions to Utility

JustinGrote committed:
Remove utilities as they have been moved into RenameService or AstExtensions

JustinGrote committed:
Rewrote Variable Handler, all tests passing except a stringconstant splat reference

JustinGrote committed:
Fix up Splat function finding, all tests pass

JustinGrote commented:
@Razmo99 got both your Visitors rewritten to `AstVisitor` and they maintain a lot less state and are more functional in processing. I also converted most of the utilities and whatnot to Ast Extension methods. There's a lot more to do with splatting that I think are workable scenarios like renaming conditional splats, but it's coming together!

Razmo99 commented:
@JustinGrote  your note for more tests for invalid renames string lead me down a rabbit hole

https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_variables?view=powershell-7.4

I didn't know that you could make variables in Powershell with spaces and other special characters. You just have to wrap them in curly brackets 

I've added a regex check for Microsoft's recommended characters, not sure we want to think about adding support for variable names with spaces in them

Razmo99 commented:
Was going to commit the following but I don't have perms on your branch for this pull request, can we merge this in?

![image](https://github.com/user-attachments/assets/826f383a-6fbf-42f6-a035-a7b2bf412c8c)

was going to add the following, regex is based on this article
https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_variables?view=powershell-7.4

```C#
        if (!Regex.IsMatch(request.NewName,@"^[\p{L}\p{Nd}_?$]+$"))
        {
            string unsupportedCharacters = "The best practice is that variable names include only alphanumeric characters and the underscore (_) character. Variable names that include spaces and other special characters, are difficult to use and should be avoided. Rename could not be completed.";
            throw new HandlerErrorException(unsupportedCharacters);
        }
```

Razmo99 commented:
> I cannot commit to your branch
> 
> ![image](https://private-user-images.githubusercontent.com/3089087/372009448-1ca065e6-5a28-483d-a95f-defa4a79622a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjc2ODU4NzIsIm5iZiI6MTcyNzY4NTU3MiwicGF0aCI6Ii8zMDg5MDg3LzM3MjAwOTQ0OC0xY2EwNjVlNi01YTI4LTQ4M2QtYTk1Zi1kZWZhNGE3OTYyMmEucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDkzMCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA5MzBUMDgzOTMyWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YjJjODE5YTRlNTYyMmE1YzY1ODU5ODViNWIxYjYwNmMwNDVmNWZlOGY0ODFiZjJhN2JiNDY3NzVkMjQ5ZjBhOCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.h1Brpi9eekvslDT9WJ19J8NLQ3lPadYvYZ4PEgRQxrI)
> 
> was going to add the following, regex is based on this article https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_variables?view=powershell-7.4
> 
> ```cs
>         if (!Regex.IsMatch(request.NewName,@"^[\p{L}\p{Nd}_?$]+$"))
>         {
>             string unsupportedCharacters = "The best practice is that variable names include only alphanumeric characters and the underscore (_) character. Variable names that include spaces and other special characters, are difficult to use and should be avoided. Rename could not be completed.";
>             throw new HandlerErrorException(unsupportedCharacters);
>         }
> ```

Just realized this should live under RenameVariable and have another regex for function chars

Razmo99 commented:
> @Razmo99 got both your Visitors rewritten to `AstVisitor` and they maintain a lot less state and are more functional in processing. I also converted most of the utilities and whatnot to Ast Extension methods. There's a lot more to do with splatting that I think are workable scenarios like renaming conditional splats, but it's coming together!

Its looking really good Justin!, its looking a lot cleaner and easier to work on, I'm learning a lot just seeing how you've reworked it

JustinGrote commented:
@Razmo99 I'll merge it in the next few days. In terms of the regex, after discussions with @seeminglyscience, we think it better to actually try to parse the variable name using the PowerShell engine and validate the results, that way if any rules change they don't have to be updated in the extension, and the general philosophy of the PS extension is to not duplicate logic that can be accessed from PowerShell itself.

JustinGrote committed:
Clean out some dead code

JustinGrote committed:
Add name validation and related tests, also rearrange/rename some test fixtures

JustinGrote committed:
Update readme with current unsupported scenarios

JustinGrote committed:
Rework function name check to be ast based

JustinGrote committed:
Add Limitation about scriptblocks

Razmo99 commented:
> @Razmo99 I'll merge it in the next few days. In terms of the regex, after discussions with @seeminglyscience, we think it better to actually try to parse the variable name using the PowerShell engine and validate the results, that way if any rules change they don't have to be updated in the extension, and the general philosophy of the PS extension is to not duplicate logic that can be accessed from PowerShell itself.

@JustinGrote now that u say it thats a much better and future proof way to design it, it would cover more edge cases aswell
